src/sync.cpp

```diff
@@ -185,7 +185,16 @@ std::string LocksHeld()
     return result;
 }
 
-void AssertLockHeldInternal(const char* pszName, const char* pszFile, int nLine, void* cs)
+void AssertLockHeldInternal(const char* pszName, const char* pszFile, int nLine, RecursiveMutex* cs)
+{
+    for (const std::pair<void*, CLockLocation>& i : g_lockstack)
+        if (i.first == cs)
+            return;
+    tfm::format(std::cerr, "Assertion failed: lock %s not held in %s:%i; locks held:\n%s", pszName, pszFile, nLine, LocksHeld());
+    abort();
+}
+
+void AssertLockHeldInternal(const char* pszName, const char* pszFile, int nLine, Mutex* cs)
```

Comment:
Right, in `C` the "any type" is `void*`, whereas in `C++` we have templates.

That `void*` typecast at the bottom of the comment above is like sweeping it under the carpet and it still leaves the possibility to use `AssertLockHeldInternal()` in a type unsafe way (because it takes `void*` argument).

Reply:
[Updated](https://github.com/bitcoin/bitcoin/pull/18635#issuecomment-626134492).